Add new `--proxy-headers` option

vmuzikar opened the pull request:
Closes #23431

Operator will be handled as a follow-up in https://github.com/keycloak/keycloak/issues/25179.

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

vmuzikar committed:
Add new `--proxy-headers` option

Closes #23431

Signed-off-by: Václav Muzikář <vmuzikar@redhat.com>

Co-authored-by: Martin Bartoš <mabartos@redhat.com>
Co-authored-by: Alexander Schwartz <aschwart@redhat.com>

shawkins commented on quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt:
```diff
@@ -150,14 +150,14 @@ HTTP(S):
 --https-trust-store-file <file>
-                     DEPRECATED: The trust store which holds the certificate information of the
-                       certificates to trust.
+                     DEPRECATED. Use: truststore-paths. -- The trust store which holds the
```
I'm not sure if it's worth it to convey, but the usage of truststore-paths would be optional as you could just use the defatult location.

vmuzikar replied:
I'll add a more descriptive note there (`deprecatedWithNote`).

vmuzikar replied:
Change the wording a bit and removed the hinted option.

shawkins commented on quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt:
```diff
@@ -150,14 +150,14 @@ HTTP(S):
+                       certificate information of the certificates to trust.
 --https-trust-store-password <password>
-                     DEPRECATED: The password of the trust store file.
+                     DEPRECATED. Use: truststore-paths. -- The password of the trust store file.
```
There will be no replacement for the password - the the store requires a password, the user will have to convert it.

vmuzikar replied:
I'll add a more descriptive note.

vmuzikar replied:
Change the wording a bit and removed the hinted option.

shawkins commented on quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt:
```diff
@@ -150,14 +150,14 @@ HTTP(S):
-                     DEPRECATED: The type of the trust store file. If not given, the type is
-                       automatically detected based on the file name. If 'fips-mode' is set to
-                       'strict' and no value is set, it defaults to 'BCFKS'.
+                     DEPRECATED. Use: truststore-paths. -- The type of the trust store file. If not
```
There is currently no replacement for type - if it's not a pkcs12 nor pem, it's not supported.

vmuzikar replied:
I'll add a more descriptive note.

vmuzikar replied:
Change the wording a bit and removed the hinted option.

vmuzikar commented:
Adding @ahus1 as a reviewer for the new deprecation logic in docs.

mabartos commented on quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt:
```diff
@@ -183,8 +183,13 @@ Metrics:
 
---proxy <mode>       The proxy address forwarding mode if the server is behind a reverse proxy.
-                       Possible values are: none, edge, reencrypt, passthrough. Default: none.
+--proxy <mode>       DEPRECATED. Use: proxy-headers. -- The proxy address forwarding mode if the
```
What about changing the format of the deprecated options to sth like this?
```
DEPRECATED (use 'proxy-headers'): The proxy ...
```

mabartos commented on services/src/main/java/org/keycloak/truststore/FileTruststoreProviderFactory.java:
```diff
@@ -77,6 +77,10 @@ public void init(Config.Scope config) {
         String configuredType = config.get("type");
 
+        if (storepath != null || pass != null || configuredType != null) {
+            log.warn("Using deprecated truststore-spi options. Consider using truststore-paths option.");
```
I'd be more explicit in this case as we already know the specific provider and format of these options. WDYT?
```suggestion
            log.warn("Using deprecated 'spi-truststore-file-*' options. Consider using 'truststore-paths' option.");
```

I'm ok with not introducing additional complexities here.

mabartos commented on docs/documentation/upgrading/topics/keycloak/changes-24_0_0.adoc:
```diff
@@ -18,3 +18,23 @@ import AuthZ from 'keycloak-js/authz';
+
+[cols="1,1"]
+|===
+|`kc.sh` (no `proxy` option set)
+|`kc.sh`
```
Is it necessary to have it here? Seems irrelevant to me in the context of `Deprecated --proxy option`. Only users who have specified the option are affected by this deprecation. WDYT?

vmuzikar replied:
I wanted to make the comparison as explicit as possible, incl. how no proxy maps to no proxy headers. Does it make any sense? :)

mabartos commented on quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/PropertyMapper.java:
```diff
@@ -96,6 +100,17 @@ ConfigValue getConfigValue(String name, ConfigSourceInterceptorContext context)
         // try to obtain the value for the property we want to map first
         ConfigValue config = convertValue(context.proceed(from));
 
+        if (config != null && getDeprecatedMetadata().isPresent()) {
+            String logMsg = String.format("Deprecated option in use: \"%s\" from %s.", option.getKey(), config.getConfigSourceName());
+            if (getDeprecatedMetadata().get().getNote() != null) {
+                logMsg += " " + getDeprecatedMetadata().get().getNote();
+            }
+            if (!getDeprecatedMetadata().get().getNewOptionsKeys().isEmpty()) {
+                logMsg += String.format(" Consider using: %s.", String.join(", ", getDeprecatedMetadata().get().getNewOptionsKeys()));
+            }
+            logger.warn(logMsg);
+        }
```
It'd be probably good to store some context about these deprecations and at the end of processing these opts just print it to a console?

pedroigor replied:
Similarly as we do for errors, right? I think it will bring more consistency, yeah ...

Perhaps as clear entry like:

```
You are using deprecated options:

* option-foo: <note>
* option-bar: <note>

These options will be removed in the feature and no longer available from future release.
```

Something better than this ...

pedroigor replied:
If we want to be really strict about it, we could even fail the build/start commands and introduce a `--use-deprecated` option to make it really painful and force people to migrate. As well as force us to take real care when designing options and deprecating them :)

vmuzikar replied:
Aggregating the log message to include all used deprecated options sounds good. Next week I'll try to figure out the way to do it.

vmuzikar replied:
Reworked that so the deprecation log message is not printed only once and aggregates all deprecated options in use.

I would not be too strict (require the users to explicitly set `--use-deprecated true`) as deprecating something should not itself be a breaking change requiring you to set a new switch.

mabartos reviewed (commented):
@vmuzikar I've just briefly gone through the PR and I put some nitpicks comments below. I'll make more deep dive review later. 

It'd be good to separate the work into 2 commits related to the `proxy-headers` and deprecation logic for better review and manageability. WDYT?

vmuzikar commented:
Thanks for the reviews!

> It'd be good to separate the work into 2 commits related to the `proxy-headers` and deprecation logic for better review and manageability. WDYT?

@mabartos Since the `proxy` option was the first major candidate for deprecation (the truststore PR was merged just yesterday), it felt natural to keep it together to actually see the deprecation logic in action. I can split it now (one PR for the deprecation logic applied to the mTLS options, and second for the proxy options) if needed. But IMHO the size of this PR is manageable so far.

ahus1 reviewed (changes requested):
@vmuzikar - thank you for these changes, they make sense to me and I understand the docs. 

I pushed one change to the table, as IMHO a table should have a heading. If you use AsciiDoc formatting in table cells, the cells should be marked with `a` which I added. 

With the new option, there are still two places where the now deprecated syntax is used in the docs. 

See:

* `reverseproxy.adoc`: section "Trust the proxy to set hostname"
* `hostname.adoc`: two usages, search for `--proxy`.

I assume all three occurrences should use the new syntax.

vmuzikar commented:
> I pushed one change to the table, as IMHO a table should have a heading. If you use AsciiDoc formatting in table cells, the cells should be marked with `a` which I added.

@ahus1 Thank you! Added a comment, not sure if the changes work as expected.

> With the new option, there are still two places where the now deprecated syntax is used in the docs.

Very good catch! Thank you.

vmuzikar commented:
@shawkins @pedroigor @mabartos @ahus1 I addressed the review comments. Feel free to give it another look when you get a chance.

vmuzikar commented:
DCO is failing due to missing signature on the accepted suggestion commit. We can override the check on merge, or I can squash the commits manually before merging.

vmuzikar commented on docs/guides/server/reverseproxy.adoc:
```diff
@@ -8,11 +8,31 @@ title="Using a reverse proxy"
+Take extra precautions to ensure that the client address is properly set by your reverse proxy via the `Forwarded` or `X-Forwarded-For` headers.
+If this header is incorrectly configured, rogue clients can set this header and trick {project_name} into thinking the client is connected from a different IP address than the actual address. This precaution can be more critical if you do any deny or allow listing of IP addresses.
+
+Some {project_name} features rely on the assumption that the remote address of the HTTP request connecting to {project_name} is the real IP address of the clients machine.
```
Removed the sentence.

ahus1 commented on docs/documentation/upgrading/topics/keycloak/changes-24_0_0.adoc:
```diff
@@ -18,3 +18,26 @@ import AuthZ from 'keycloak-js/authz';
+
+The `--proxy` option has been deprecated and will be removed in a future release. The following table explains how the deprecated option maps to supported options.
+
+[%autowidth,cols="a,a"]
+|===
+| Deprecated usage | New usage
```
I ran the following command: 

```
mvn clean install -am -pl docs/documentation/upgrading -Pdocumentation
``` 

This outputs a file-link on the console. Once I click on it, I see the following which is with the headline: 

![image](https://github.com/keycloak/keycloak/assets/3957921/9ca5a539-bd72-43bb-b7f3-c41d9a9adcac)

vmuzikar replied:
Thanks! Today, I was forgetting the `-Pdocumentation`. (Last time when I wrote the docs I included it.)

ahus1 commented on docs/guides/server/hostname.adoc:
```diff
@@ -133,7 +133,7 @@ In this example, the server is accessible using a port other than the default po
 In this example, the server is running behind a proxy and both the server and the proxy are using their own certificates, so the communication between {project_name} and the proxy is encrypted. Because we want the proxy to use its own certificate, the proxy mode `reencrypt` will be used. We need to keep in mind that the proxy configuration options (as well as hostname configuration options) are n ...
```
This is also referring to proxy mode `reencrypt`, and should be updated.

ahus1 commented on docs/guides/server/hostname.adoc:
```diff
@@ -133,7 +133,7 @@ In this example, the server is accessible using a port other than the default po
-<@kc.start parameters="--proxy=reencrypt --https-port=8543 --hostname-url=https://my-keycloak.org:8443 --hostname-admin-url=https://admin.my-keycloak.org:8443"/>
+<@kc.start parameters="--proxy-headers=forwarded|xforwarded --https-port=8543 --hostname-url=https://my-keycloak.org:8443 --hostname-admin-url=https://admin.my-keycloak.org:8443"/>
 
 Note: there is currently no difference between the `passthrough` and `reencrypt` modes. For now, this is meant for future-proof configuration compatibility. The only difference is that when the `edge` proxy mode is used, HTTP is implicitly enabled (again as mentioned above, this does not affect the server behaviour).
```
This paragraph needs to be changed to match the updated CLI above, as it is still talking about `passthrough` and `reencrypt`.

ahus1 reviewed (changes requested):
Sorry, found two paragraph before/after the CLI which should also change - probably they can be a lot shorter now?

ahus1 reviewed (dismissed):
Thank you, looks good!

mabartos commented on docs/guides/templates/options.adoc:
```diff
@@ -22,6 +22,12 @@
 *Env:* `${option.keyEnv}`
 --
+
+<#if option.deprecated?has_content>
+*DEPRECATED.*
+${option.deprecated.note!}<#if option.deprecated.newOptionsKeys?has_content><#if option.deprecated.note?has_content> </#if>Use: <#list option.deprecated.newOptionsKeys as key>`+${key}+`<#if key?has_next>, </#if></#list>.</#if>
+</#if>
+
```
Works as expected.

![image](https://github.com/keycloak/keycloak/assets/38039883/bd9ebf9e-e55c-46d6-aa58-b62678b8e523)

mabartos commented on quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/cli/Picocli.java:
```diff
@@ -307,21 +313,48 @@ public static void validateNonCliConfig(List<String> cliArgs, AbstractCommand ab
+        }
+
+        if (!deprecatedInUse.isEmpty()) {
+            logger.warn("The following used options are DEPRECATED and will be removed in a future release:\n" + String.join("\n", deprecatedInUse));
```
Works as expected (provided `--cache` and `--https-trust-store-password`).

Small improvement would be to get rid of the KC namespace as we don't deprecate any other props. For property config file source the current state is relevant, but for CLI not so much. But, it's only a nitpicking.   
```
2023-12-11 12:10:03,314 WARN  [org.keycloak.quarkus.runtime.cli.Picocli] (main) The following used options are DEPRECATED and will be removed in a future release:
        - kc.https-trust-store-password: Use the System Truststore instead, see the docs for details.
        - kc.proxy: Consider using proxy-headers.
```

pedroigor commented on quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt:
```diff
@@ -150,14 +150,17 @@ HTTP(S):
 --https-trust-store-file <file>
-                     DEPRECATED: The trust store which holds the certificate information of the
-                       certificates to trust.
+                     DEPRECATED. Use the System Truststore instead, see the docs for details. --
```
Is it expected the `--` here?

vmuzikar replied:
Rewrote the message a bit. Now it looks like
<img width="876" alt="Screenshot 2023-12-11 at 15 08 00" src="https://github.com/keycloak/keycloak/assets/13906855/66a3cfab-fbe8-46c3-96fd-25e5b7321340">
<img width="876" alt="Screenshot 2023-12-11 at 14 37 03" src="https://github.com/keycloak/keycloak/assets/13906855/a89469af-c38e-49f2-ba1d-c08beeb4e9b7">

pedroigor commented on quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.unix.approved.txt:
```diff
@@ -150,14 +150,17 @@ HTTP(S):
+                       The trust store which holds the certificate information of the certificates
+                       to trust.
 --https-trust-store-password <password>
-                     DEPRECATED: The password of the trust store file.
```
I think the option description is a bit confusing. Perhaps:

```
DEPRECATED: The password of the trust store file.  Use the System Truststore instead, see the docs for details..
```

vmuzikar replied:
Rewrote the message a bit.

mabartos reviewed (dismissed):
@vmuzikar Nice work! LGTM.

pedroigor commented on quarkus/config-api/src/main/java/org/keycloak/config/OptionBuilder.java:
```diff
@@ -111,12 +114,32 @@ public List<String> get() {
+        return this;
+    }
+
+    public OptionBuilder<T> deprecatedWithNote(String note) {
```
Isn't enough the `deprecatedWithNote(String, String...)` method?

Also can we just name it `deprecated`? I mean, `WithNote` seems redundant as the parameter name is kinda making clear about setting a `note`.

To make easier to set deprecation without a note but the new options we just use a `deprecate(Set<Option>)`. So you have:

```
builder.deprecate(Set.of(OptionA, OptionB));
```

vmuzikar replied:
Done.

pedroigor commented on docs/guides/server/hostname.adoc:
```diff
@@ -40,7 +40,7 @@ Most of the time, it should be enough to set the `hostname` option in order to c
 
 * `https` scheme is used unless you set `hostname-strict-https=false`
-* if the `proxy` option is set, the proxy will use the default ports (i.e.: 80 and 443).  If the proxy uses a different port, it needs to be specified via the `hostname-port` configuration option
+* if the `proxy-headers` option is set, the proxy will use the default ports (i.e.: 80 and 443).  If the proxy uses a different port, it needs to be specified via the `hostname-port` configuration option
```
IMO, we should not recommend using `hostname-port` but using the `hostname-url` option.

This option does not much sense if you can just provide the URL you want to expose the frontend URLs (same thing for admin). The `hostname-port` is kinda another option causing the confusion you want to solve with the on-going changes in the hostname configuration and implementation.

vmuzikar replied:
Done.

vmuzikar committed:
Address review comments vol. 03

Signed-off-by: Václav Muzikář <vmuzikar@redhat.com>

vmuzikar committed:
Address review comments vol. 04

Signed-off-by: Václav Muzikář <vmuzikar@redhat.com>

mabartos reviewed (dismissed):
@vmuzikar LGTM

mabartos commented:
@vmuzikar Oops, conflict.

vmuzikar commented:
> Oops, conflict.

Rebased, thanks.

mabartos committed:
Merge branch 'main' into proxy-headers

mabartos commented:
@vmuzikar @pedroigor Resolved conflicts and rebased.
